Revert if not enough gas

anxolin opened the pull request:
Makes trampoline contract to revert if there's no enough gas left

## Context
Solvers might under-estimate the gas required for a settlement with hooks.

Hooks have a `gasLimit` that they require, however, the settlement transaction can indicate a low gas, so when we arrive to the hook the gas left is less than the `gasLimit`. 
This will likely result in the hook not getting some fair gas to execute and failing. 

This goes unnoticed for the solver, because the settlement will succeed. 

This PR makes sure the gas left in the transaction is enough to honor the instructed `gasLimit`, if not, it reverts. This would make it so the solvers get a revert in the settlement simulation if this is the case.


## Changes
Before each hook execution, we check there's enough gas left:

`require(gasleft() >= hook.gasLimit, "Not enough gas");`

## Unit tests
Added 2 tests:
- `test_RevertsWhenNotEnoughGas`
- `test_RevertsWhenNotEnoughGasForMultipleHooks`


## Test
```
forge test
```

anxolin committed:
Revert if not enough gas

carlos-cow commented on test/HooksTrampoline.t.sol:
```diff
@@ -160,3 +198,11 @@ contract Hummer {
+
+contract BurnGas {
+    function consumeAllGas() public {
+        while (true) {
```
does this finish fast enough?

You can also try to replace with this https://github.com/Vectorized/solady/blob/main/src/utils/GasBurnerLib.sol#L49
for faster gas burn

carlos-cow replied:
apparently it runs fast enough, nevermind
```
Ran 1 test suite in 6.59ms (1.19ms CPU time): 8 tests passed, 0 failed, 0 skipped (8 total tests)
```

anxolin committed:
Add verification

anxolin committed:
Declare error

anxolin committed:
Use hummer contract instead of gas burner

anxolin commented on test/HooksTrampoline.t.sol:
```diff
@@ -160,3 +198,11 @@ contract Hummer {
+            // Do nothing, just consume gas
+        }
+    }
+}
```
9b8e2b02f8e30fa34e773efd26fc56a1f4772440

anxolin committed:
Clarify comment

Co-authored-by: Federico Giacon <58218759+fedgiac@users.noreply.github.com>

anxolin committed:
Clarify comment

Co-authored-by: Federico Giacon <58218759+fedgiac@users.noreply.github.com>

anxolin committed:
Isolate tests

anxolin committed:
Add gas limit enforcement tests

anxolin committed:
Demostrate 1/64 reservation

anxolin commented on src/HooksTrampoline.sol:
```diff
@@ -55,6 +55,7 @@ contract HooksTrampoline {
             Hook calldata hook;
             for (uint256 i; i < hooks.length; ++i) {
                 hook = hooks[i];
+                require(gasleft() >= hook.gasLimit, "Not enough gas");
```
I've made the test to be isolate=true (except one that was breaking).

Not sure if this is what you would want, but we could do something to 
```solidity
uint256 effectiveGas = gasleft() * 63 / 64;
if (effectiveGas < hook.gasLimit) {
    revert NotEnoughGas();
}
```

Yes, I guess your concern is valid. Someone that want to not break the rules, could technically craft a transaction that leaves just too little gas, this 1/64 eams my tests above passes, but the hook does't get enough gas (compared to the limit).

However, is this really so bad and worth it to overcomplicate the logic? I would think, as long as hooks gas are over-estimated (by 1.57%) then its not possible to cause this. 

Assuming I'm getting it, I would be inclined to go with this version tomorrow live

anxolin commented:
@fedgiac @carlos-cow  I addressed most of the comments

Also, I added all the tests you suggested in the chat under `GasLimitEnforcementTest` (I just broke them down and refactored them). 

I also tested that indeed 1/64 of the remainder is not forwarded. See `test_DemonstratesOneSixtyFourthGasReservation` . I did not create a test that reproduces the edge case, if important I can try to reproduce tomorrow. I would also have some time tomorrow to review more failed transactions before swap and bridge goes live.

fedgiac committed:
Account for unsent 1/64th of gas in call

fedgiac reviewed (commented):
My comments are in https://github.com/cowprotocol/hooks-trampoline/pull/15.

fleupold commented on src/HooksTrampoline.sol:
```diff
@@ -55,6 +63,12 @@ contract HooksTrampoline {
+                // math is used as a heuristic to account for this.
+                uint256 forwardedGas = gasleft() * 63 / 64;
+                if (forwardedGas < hook.gasLimit) {
+                    revert NotEnoughGas();
```
I'm afraid this implementation suffers from breaking the default `eth_estimateGas` implementation of many RPC nodes. Given this is a vanilla revert (not a real running out of gas) RPCs don't know that they can avoid this by increasing the gas estimate.

One way to repro is to try to estimate gas for a hook that uses a gasLimit of 10M (it will just say that it reverts). Do you think you can come up with an implementation where a hook with this gas limit can pass the `estimateGas` call for most nodes (reth, geth, erigon)?

anxolin committed:
Change license

Co-authored-by: Federico Giacon <58218759+fedgiac@users.noreply.github.com>

fedgiac committed:
If gas is missing, revert by wasting gas

anxolin committed:
Add verifier address

anxolin committed:
Fix error in command

anxolin commented:
@fedgiac can I mege? I want to mege this to main, and then https://github.com/cowprotocol/hooks-trampoline/pull/18

fedgiac reviewed (dismissed):
Looks good! (Comments are just nits, feel free to ignore.)

anxolin committed:
Use absolute path

Co-authored-by: Federico Giacon <58218759+fedgiac@users.noreply.github.com>
